### PR TITLE
initial login race condition fix

### LINKED ISSUE #244: Sync is not working on initial sign in
## Steps to reproduce

1. Open Lockbox on any device
2. Tap on "Get Started"
3. Login with a valid username and password.
4. Wait for sync

### Expected behaviour
Sync is performed and a list of logins should populate the "All entries" screen

### Actual behaviour
No sync is performed

### Device and build information
- Nexus 5 (Android 6.0.1)
- HTC Desire 820 (Android 6.0.1)
- Nokia 6 (Android 7.1.1)
- OnePlus 5T (Android 8.1.0)
- Google Pixel (Android 9)
- [Build #1316](https://app.bitrise.io/build/67245d12e783e79d)

### Notes

- Nothing was synced even after 20 minutes of waiting
- [Logcat](https://github.com/mozilla-lockbox/lockbox-android/files/2624537/log.txt)
- screen

### PR DESCRIPTION
**Pull request by sashei**
Fixes #244 

## To Do

- [x] make sure CI builds are passing (e.g.: fix lint and other errors)

### CODE REVIEW

**Comment by sashei** [comment]
note: we still see an ugly fxa redirect screen before going to the item list with this solution. haven't figured out a fix for that yet.

File: app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt [hunk 1]
```diff
@@ -36,9 +35,6 @@ class FxALoginPresenter(
             url?.let {
                 if (url.startsWith(Constant.FxA.redirectUri)) {
                     dispatcher.dispatch(AccountAction.OauthRedirect(url))
-                    // TODO: remove following line when centralized routing is implemented via:
-                    // https://github.com/mozilla-lockbox/lockbox-android/issues/144
-                    dispatcher.dispatch(DataStoreAction.Unlock)
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt]: 😱

File: app/src/main/java/mozilla/lockbox/store/DataStore.kt [hunk 2]
```diff
@@ -190,18 +192,17 @@ open class DataStore(
...
-            sync()
-        }
-
         if (backend.isLocked()) {
+            stateSubject.accept(State.Unlocking)
             backend.unlock(support.encryptionKey)
                 .asSingle(Dispatchers.Default)
-                .subscribe { _, _ -> initialSync() }
+                .subscribe { _, _ ->
+                    stateSubject.accept(State.Unlocked)
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/DataStore.kt]: Putting this here makes much more sense.

File: app/src/main/java/mozilla/lockbox/store/DataStore.kt [hunk 3]
```diff
@@ -190,18 +192,17 @@ open class DataStore(
             return
         }
 
-        fun initialSync() {
-            unlock()
-            sync()
-        }
-
         if (backend.isLocked()) {
+            stateSubject.accept(State.Unlocking)
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/store/DataStore.kt]: I think we should get rid of the `.Unlocking` state.

File: app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt [hunk 4]
```diff
@@ -79,7 +81,7 @@ class RoutePresenter(
 
     private fun dataStoreToRouteActions(storageState: State): Optional<RouteAction> {
         return when (storageState) {
-            is State.Unlocking -> null
+            is State.Unlocking -> RouteAction.ItemList
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt]: I'd like to kill off `.Unlocking`.

File: app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt (hunk 1 above)
Comment (linuxwolf) [reply app/src/main/java/mozilla/lockbox/presenter/FxALoginPresenter.kt]: I did indeed LOL

File: app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt (hunk 4 above)
Comment (sashei) [reply app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt]: _done_

**Comment by devinreams** [comment]
>I wanted to test this with an acct with lots of logins (which you are presumably doing?)

Yes, both firefoxlockbox@gmail.com and dreams@mozilla.com

>I can file a follow-up bug, or just amend the first onboarding issue to note that we should look for it there. Thoughts?

I agree that tackling it there in that issue is fine. Do you mind amending/adding the note to that issue so its meaningful and sparks our memory? Thank you very much! 🙇

**Comment by sashei** [comment]
@devinreams @linuxwolf _done_. although I realized that depending on the user's hardware / device version, it's possible that any of the three onboarding screens will be the first one they see 😆 so I've amended all of them.

**Pull request merged**